[clang] Remove FEM_Indeterminable

ojhunt opened the pull request:
Remove FEM_Indeterminable as it is unused and cannot be stored safely in an unsigned bitfield

ojhunt committed:
[clang] Remove FEM_Indeterminable

Remove FEM_Indeterminable as it is unused and cannot be stored safely in an unsigned bitfield

ojhunt commented:
(Recreating after mismerge)

ojhunt commented:
@Endilll this is the fix for the warning you're seeing

ojhunt commented:
@zahiraam we started diagnosing size/type mismatch of bitfields vs preferred_type in, which exposed that FPEvalMethodKind based options can't store FEM_Indeterminable https://github.com/llvm/llvm-project/pull/116785

ojhunt commented:
@zahiraam @AaronBallman a different option would be to add a signed vs unsigned storage option to the `OPTION` and `BENIGN_ENUM_LANGOPT` macros so we can store negative enumerations safely

zahiraam commented:
> @zahiraam @AaronBallman a different option would be to add a signed vs unsigned storage option to the `OPTION` and `BENIGN_ENUM_LANGOPT` macros so we can store negative enumerations safely

I think I would prefer this solution. We need to be able to set the evaluation method to a value (-1) when it can't be known from the target or when the value of `ffp-eval-method`  is inconsistent with the target.

AaronBallman commented:
> > @zahiraam @AaronBallman a different option would be to add a signed vs unsigned storage option to the `OPTION` and `BENIGN_ENUM_LANGOPT` macros so we can store negative enumerations safely
> 
> I think I would prefer this solution. We need to be able to set the evaluation method to a value (-1) when it can't be known from the target or when the value of `ffp-eval-method` is inconsistent with the target.

Could we shift all the values, so `FEM_Indeterminable` is `0`?

zahiraam commented:
> > > @zahiraam @AaronBallman a different option would be to add a signed vs unsigned storage option to the `OPTION` and `BENIGN_ENUM_LANGOPT` macros so we can store negative enumerations safely
> > 
> > 
> > I think I would prefer this solution. We need to be able to set the evaluation method to a value (-1) when it can't be known from the target or when the value of `ffp-eval-method` is inconsistent with the target.
> 
> Could we shift all the values, so `FEM_Indeterminable` is `0`?

I don't think so. In practice on Linux systems, they use `__FLT_EVAL_METHOD__` to control the type of `float_t` and `double_t`. Things like this:

`# if __FLT_EVAL_METHOD__ == -1`
`#  define __GLIBC_FLT_EVAL_METHOD       2`
....
```
# if __GLIBC_FLT_EVAL_METHOD == 0 || __GLIBC_FLT_EVAL_METHOD == 16
typedef float float_t;
typedef double double_t;
# elif __GLIBC_FLT_EVAL_METHOD == 1
typedef double float_t;
typedef double double_t;
# elif __GLIBC_FLT_EVAL_METHOD == 2
typedef long double float_t;
typedef long double double_t;
# elif __GLIBC_FLT_EVAL_METHOD == 32
typedef _Float32 float_t;
typedef double double_t;
...
```

ojhunt commented:
Ah, slot FLT_EVAL_METHOD is used in source as well, in that case I think the fix is to make signed options an option (pun intended)

ojhunt commented:
although, what happens currently? Per the warning we should be producing the wrong value if there's a direct enum value to code visible value mapping

AaronBallman commented:
> > > > @zahiraam @AaronBallman a different option would be to add a signed vs unsigned storage option to the `OPTION` and `BENIGN_ENUM_LANGOPT` macros so we can store negative enumerations safely
> > > 
> > > 
> > > I think I would prefer this solution. We need to be able to set the evaluation method to a value (-1) when it can't be known from the target or when the value of `ffp-eval-method` is inconsistent with the target.
> > 
> > 
> > Could we shift all the values, so `FEM_Indeterminable` is `0`?
> 
> I don't think so. In practice on Linux systems, they use `__FLT_EVAL_METHOD__` to control the type of `float_t` and `double_t`. Things like this:

Oh shoot, I forgot this was tied to `__FLT_EVAL_METHOD__`

ojhunt commented:
> > > > > @zahiraam @AaronBallman a different option would be to add a signed vs unsigned storage option to the `OPTION` and `BENIGN_ENUM_LANGOPT` macros so we can store negative enumerations safely
> > > > 
> > > > 
> > > > I think I would prefer this solution. We need to be able to set the evaluation method to a value (-1) when it can't be known from the target or when the value of `ffp-eval-method` is inconsistent with the target.
> > > 
> > > 
> > > Could we shift all the values, so `FEM_Indeterminable` is `0`?
> > 
> > 
> > I don't think so. In practice on Linux systems, they use `__FLT_EVAL_METHOD__` to control the type of `float_t` and `double_t`. Things like this:
> 
> Oh shoot, I forgot this was tied to `__FLT_EVAL_METHOD__`

As far as I can make out this isn't an automatic mapping, so we can always just set it correctly, but more importantly, I cannot see how it is not currently broken - loading FEM_Indeterminate should not (in principle) be sign extending.

How do I get clang into the FEM_Indeterminate mode? As far as I can make out, we currently have a single test for `__FLT_EVAL_METHOD__ == -1` and it does not get run/checked

zahiraam commented:
> > > > > > @zahiraam @AaronBallman a different option would be to add a signed vs unsigned storage option to the `OPTION` and `BENIGN_ENUM_LANGOPT` macros so we can store negative enumerations safely
> > > > > 
> > > > > 
> > > > > I think I would prefer this solution. We need to be able to set the evaluation method to a value (-1) when it can't be known from the target or when the value of `ffp-eval-method` is inconsistent with the target.
> > > > 
> > > > 
> > > > Could we shift all the values, so `FEM_Indeterminable` is `0`?
> > > 
> > > 
> > > I don't think so. In practice on Linux systems, they use `__FLT_EVAL_METHOD__` to control the type of `float_t` and `double_t`. Things like this:
> > 
> > 
> > Oh shoot, I forgot this was tied to `__FLT_EVAL_METHOD__`
> 
> As far as I can make out this isn't an automatic mapping, so we can always just set it correctly, but more importantly, I cannot see how it is not currently broken - loading FEM_Indeterminate should not (in principle) be sign extending.
> 
> How do I get clang into the FEM_Indeterminate mode? As far as I can make out, we currently have a single test for `__FLT_EVAL_METHOD__ == -1` and it does not get run/checked

See issue https://github.com/llvm/llvm-project/issues/137600

ojhunt commented:
@zahiraam none of these actually try to set FEM_Indeterminate, so far I cannot find any way to make clang ever set that state and I can’t find a way for source code to do so either - it looks like in the past the fpcontrol pragma could induce it, but none of the examples still do so in clang _or_ gcc.

I really think the correct thing to do is to remove FEM_Indeterminate because even if you could currently set the clang fp eval mode it would not be presented as -1 in __FLT_EVAL_METHOD__ because of the problem this warning is complaining about (ie clang has not been able to produce __FLT_EVAL_METHOD__==-1 in a long time)

ojhunt commented:
@AaronBallman @zahiraam  given we do not support FEM_Indeterminate, and there appears to be no way to actually get clang into the state anyway, I think that the correct thing to do is to remove it and work out had to get this to cooperate with the OPTION macro if we ever do add support for it, does this seem reasonable?

ojhunt commented:
Yeah, clang-15 seems to be the only time we would ever produce -1: https://godbolt.org/z/7M61eff4f

AaronBallman reviewed (approved):
I think this is a reasonable way forward for right now; we can bring back the enumerator once we're actually using it. WDYT @zahiraam ?

zahiraam commented:
> I think this is a reasonable way forward for right now; we can bring back the enumerator once we're actually using it. WDYT @zahiraam ?

I tend to think so too. But I am waiting for a response from the math library people internally to make sure that we don't loose compatibility with `icc` doing this.

ojhunt commented:
> > I think this is a reasonable way forward for right now; we can bring back the enumerator once we're actually using it. WDYT @zahiraam ?
> 
> I tend to think so too. But I am waiting for a response from the math library people internally to make sure that we don't loose compatibility with `icc` doing this.

@zahiraam @AaronBallman r+'d this, and it's causing a bunch of warning noise for folk, so I'd like to land this soon - do you have any idea when icc folk can get back to you? Also, if they can find an issue can you find out how they're getting clang into the FEM_Indeterminate state? I can't find any way to trigger it, and as far as I can make out even if we could I believe we would currently set `__FLT_EVAL_METHOD__` to 7 so I'm really not sure how that could break anything

zahiraam commented:
@ojhunt Please go ahead and merge this in. Thanks for your patience.

zahiraam reviewed (approved):
Looks good. Thanks.

zahiraam commented:
@ojhunt  In https://github.com/llvm/llvm-project/issues/137600 there are a list of a few temporary PRs that were merged in to silence the warning. These need to be reverted after this one gets merged in. Do you want to revert them or you want me to do that?

ojhunt committed:
Revert c2a62af

ojhunt commented:
Just waiting on a build (force push was due to me hopelessly screwing up my tree)

ojhunt commented:
> @ojhunt In #137600 there are a list of a few temporary PRs that were merged in to silence the warning. These need to be reverted after this one gets merged in. Do you want to revert them or you want me to do that?

I removed them as part of resolving the merge conflict, and reverted https://github.com/swiftlang/llvm-project/commit/c2a62af2a51d58183bcd72ee8a86c37ddd526758 manually

zahiraam commented:
@ojhunt I think there a more than one revert. Can you please read from https://github.com/llvm/llvm-project/issues/137600#issuecomment-2842037532 to see all the reverts that need to be done. Thanks.

ojhunt commented:
Two of the fixes were to code that was removed while resolving the merge conflict on the LANG_OPT declaration, I thought with the revert of c2a62af I got all the changes but I'll recheck

zahiraam commented:
> Two of the fixes were to code that was removed while resolving the merge conflict on the LANG_OPT declaration, I thought with the revert of [c2a62af](https://github.com/llvm/llvm-project/commit/c2a62af2a51d58183bcd72ee8a86c37ddd526758) I got all the changes but I'll recheck

Thanks.

marius-cornea commented:
An after-the-fact note: for Intel math libraries, this does not matter at all, as we never use FLT_EVAL_METHOD in our code. We use however other methods to achieve the effect we want, i.e., the equivalent of FLT_EVAL_METHOD = 0 (source). We use fp-model precise, or sometimes strict for that.
